src/blocks/disk_iostats.rs

```diff
@@ -0,0 +1,169 @@
+            }
+            Some(device) => {
+                let mut widget = Widget::new();
+
+                widget.set_format(format.clone());
+
+                let new_stats = read_stats(&device).await?;
+                let sector_size = read_sector_size(&device).await?;
+
+                let mut speed_read = 0.0;
+                let mut speed_write = 0.0;
+                if let Some(old_stats) = old_stats {
+                    let diff = new_stats - old_stats;
+                    let elapsed = stats_timer.elapsed().as_secs_f64();
+                    stats_timer = Instant::now();
+                    let size_read = diff.sectors_read as u64 * sector_size;
+                    let size_written = diff.sectors_written as u64 * sector_size;
+                    speed_read = size_read as f64 / elapsed;
+                    speed_write = size_written as f64 / elapsed;
+                };
+                old_stats = Some(new_stats);
+
+                widget.set_values(map! {
+                    "icon" => Value::icon("disk_drive"),
+                    "speed_read" => Value::bytes(speed_read),
+                    "speed_write" => Value::bytes(speed_write),
+                    "device" => Value::text(device),
+                });
+
+                api.set_widget(widget)?;
+            }
+        }
+
+        select! {
+            _ = timer.tick() => continue,
+            _ = api.wait_for_update_request() => continue,
+        }
+    }
+}
+
+async fn find_device() -> Result<Option<String>> {
+    let mut sysfs_dir = read_dir(BLOCK_DEVICES_PATH)
+        .await
+        .error("Failed to open /sys/class/block directory")?;
+    while let Some(dir) = sysfs_dir
+        .next_entry()
+        .await
+        .error("Failed to read /sys/class/block directory")?
+    {
+        let path = dir.path();
+        if path.join("device").exists() {
+            return Ok(Some(
+                dir.file_name()
+                    .into_string()
+                    .map_err(|_| Error::new("Invalid device filename"))?,
+            ));
+        }
+    }
+
+    Ok(None)
+}
+
+#[derive(Debug, Default, Clone, Copy)]
+struct Stats {
+    sectors_read: c_ulong,
+    sectors_written: c_ulong,
+}
+
+impl ops::Sub for Stats {
+    type Output = Self;
+
+    fn sub(mut self, rhs: Self) -> Self::Output {
+        self.sectors_read = self.sectors_read.wrapping_sub(rhs.sectors_read);
+        self.sectors_written = self.sectors_written.wrapping_sub(rhs.sectors_written);
+        self
+    }
+}
+
+async fn read_stats(device: &str) -> Result<Stats> {
+    let raw = read_file(Path::new(BLOCK_DEVICES_PATH).join(device).join("stat"))
+        .await
+        .error("Failed to read stat file")?;
+    let fields: Vec<&str> = raw.split_whitespace().collect();
+    Ok(Stats {
+        sectors_read: fields[2].parse().error("Failed to parse sectors read")?,
+        sectors_written: fields[6].parse().error("Failed to parse sectors written")?,
```

Comment:
Let's use `fields.get(n).error("bad stat file")?` in case for some reason there are not enough fields

Reply:
Done